File: judgels-legacy/jerahmeel/app/org/iatoki/judgels/jerahmeel/course/chapter/CourseChapter.java
package org.iatoki.judgels.jerahmeel.course.chapter;

public final class CourseChapter {

    private final long id;
    private final String courseJid;
    private final String chapterJid;
    private final String alias;

    public CourseChapter(long id, String courseJid, String chapterJid, String alias) {
        this.id = id;
        this.courseJid = courseJid;
        this.chapterJid = chapterJid;
        this.alias = alias;
    }

    public long getId() {
        return id;
    }

    public String getCourseJid() {
        return courseJid;
    }

    public String getChapterJid() {
        return chapterJid;
    }

    public String getAlias() {
        return alias;
    }
}

File: judgels-frontends/raphael/src/modules/api/jerahmeel/courseChapter.ts
import { get, put } from '../../../modules/api/http';

import { ChapterInfo, Chapter, ChapterProgress } from './chapter';
import { baseCourseURL } from './course';

export interface CourseChapter {
  alias: string;
  chapterJid: string;
}

export interface CourseChaptersResponse {
  data: CourseChapter[];
  chaptersMap: { [chapterJid: string]: ChapterInfo };
  chapterProgressesMap: { [chapterJid: string]: ChapterProgress };
}

const baseURL = (courseJid: string) => `${baseCourseURL(courseJid)}/chapters`;

export const courseChapterAPI = {
  getChapters: (token: string, courseJid: string): Promise<CourseChaptersResponse> => {
    return get(baseURL(courseJid), token);
  },

  setChapters: (token: string, courseJid: string, data: CourseChapter[]): Promise<void> => {
    return put(baseURL(courseJid), token, data);
  },

  getChapter: (token: string, courseJid: string, chapterAlias: string): Promise<Chapter> => {
    return get(`${baseURL(courseJid)}/${chapterAlias}`, token);
  },
};
